src/audio_pixel.cpp

```diff
@@ -0,0 +1,270 @@
+
+                // set max to equal reaper min, and min to equal reaper max
+                curr_pixel = audio_pixel_t();
+                collected_samples = 0;
+                pixels_per_channel++;
+            }
+        }
+    }
+}
+
+std::vector<std::vector<audio_pixel_t>> audio_pixel_block_t::get_pixels(double t0, double t1) {
+    std::vector<std::vector<audio_pixel_t>> output_block;
+    int starting_idx =  std::max(0, (int)(t0 * m_pix_per_s));
+    int ending_idx = std::min((int)(t1 * m_pix_per_s), m_block_size);
+
+    for (int channel = 0; channel < m_channel_pixels.size(); channel++){
+        std::vector<audio_pixel_t> channel_block = m_channel_pixels[channel];
+        std::vector<audio_pixel_t> subset_block;
+        for (int i = starting_idx; i < ending_idx && i < channel_block.size(); i++) {
+            subset_block.push_back(channel_block[i]);
+        }
+        output_block.push_back(subset_block);
+    }
+
+    return output_block;
+}
+
+// ************* audio_pixel_mipmap_t ****************
+
+audio_pixel_block_t audio_pixel_mipmap_t::get_pixels(double t0, double t1, int pix_per_s) {
+    int nearest_pps = get_nearest_pps(pix_per_s);
+
+    if (nearest_pps == pix_per_s){ // last edge case, when the given pps is already a block (other cases in get nearest pps)
+        audio_pixel_block_t matching_block = m_blocks[nearest_pps];
+        audio_pixel_block_t output_block(pix_per_s, matching_block.get_pixels(t0, t1));
+        return output_block;
+    }
+
+    // perform interpolation
+    audio_pixel_block_t interpolated_block = create_interpolated_block(nearest_pps, pix_per_s, t0, t1);
+    return interpolated_block;
+};
+
+audio_pixel_block_t audio_pixel_mipmap_t::get_block(int pix_per_s){
+    int nearest_pps = get_nearest_pps(pix_per_s);
+
+    if (nearest_pps == pix_per_s){
+        audio_pixel_block_t matching_block = m_blocks[nearest_pps];
+        return matching_block;
+    }
+
+    // perform interpolation
+    audio_pixel_block_t interpolated_block = create_interpolated_block(nearest_pps, pix_per_s, -1, -1);
+    return interpolated_block;
+}
+
+int  audio_pixel_mipmap_t::get_nearest_pps(int pix_per_s) {
+    int n_blocks = m_block_pps.size();
+
+    if (pix_per_s <= m_block_pps[0]) // given pps smaller than our smallest rep
+        return m_block_pps[0];
+    else if (pix_per_s >= m_block_pps[n_blocks - 1]) // given pps larger than our largest rep
+        return m_block_pps[n_blocks - 1];
+
+    return get_nearest_pps_helper(pix_per_s);
+}
+
+audio_pixel_block_t audio_pixel_mipmap_t::create_interpolated_block(int src_pps, int new_pps, double t0, double t1) {
+    if (t0 == -1)
+        t0 = GetAudioAccessorStartTime(m_accessor.get());
+    if (t1 == -1)
+        t1 = GetAudioAccessorEndTime(m_accessor.get());
+
+    // interpolate from the block with nearest-pps to create a new block at new_pps
+    auto map_entry = m_blocks.find(src_pps);
+    audio_pixel_block_t nearest_block = map_entry->second;
```

Comment:
this is making an unnecessary copy of the block. maybe we should make the copy constructor of `audio_pixel_block_t` private so we don't do any unnecessary copying?

Reply:
I cannot make it private any tips? but used that to prevent any calls to copy constructor